README.md
<p align="center">

<img src="https://github.com/jina-ai/now/blob/main/docs/_static/logo-light.svg?raw=true" alt="Jina NOW logo: The data structure for unstructured data" width="300px">  


<br>
One command to host them all. Bring your search case into the cloud in minutes. <br>
Tell us what you think: <a href="https://10sw1tcpld4.typeform.com/to/VTAyYRpR?utm_source=cli">Survey</a>
</p>

<p align=center>
<a href="https://pypi.org/project/jina-now/"><img src="https://github.com/jina-ai/jina/blob/master/.github/badges/python-badge.svg?raw=true" alt="Python 3.7 3.8 3.9 3.10" title="Jina NOW supports Python 3.7 and above"></a>
<a href="https://pypi.org/project/jina-now/"><img src="https://img.shields.io/pypi/v/jina-now?color=%23099cec&amp;label=PyPI&amp;logo=pypi&amp;logoColor=white" alt="PyPI"></a>
</p>


<p align="center">
<img src="https://user-images.githubusercontent.com/11627845/164569398-5ef22a41-e2e1-438a-88a5-2ac43ad9426d.gif" alt="Jina NOW logo: The data structure for unstructured data" width="600px">

<!-- start elevator-pitch -->

NOW gives the world access to neural image search with just one command.
 
- ⛅ **Cloud**: We handle deployment and maintenance.
- 🐎 **Fast and easy**: Set up your search use case in minutes with minimal effort.
- 🌈 **Quality**: You provide the labels, NOW fine-tunes the model.
- ✨ **Nocode**: Non-technical people can deploy with ease.

Read how [Jina NOW is production ready](https://now.jina.ai/user-guides/production/).

### Install

```bash
pip install jina-now
```

If you need sudo to run Docker, use sudo to install and use Jina NOW as well.

#### Mac M1

For the M1 we recommend using a [Conda environment](https://docs.jina.ai/get-started/install/troubleshooting/#on-mac-m1).
In a new Conda environment:

1. Run `conda install grpcio tokenizers protobuf`
2. Run `pip install jina-now`.

### Quick start
```bash
jina now start
```
**1.** Choose your data source. 

You may either choose a demo dataset hosted by NOW, or use your own custom dataset, to build an application.
NOW can support your custom data in the form of a `DocumentArray`, as a path to a local folder, or S3 bucket.

```commandline
? How do you want to provide input? (format: https://docarray.jina.ai/)  (Use arrow keys)
 ❯ Demo dataset
   DocumentArray name (recommended)
   Local folder
   S3 bucket
   - Elasticsearch (will be available in upcoming versions)
```

You can choose a demo dataset to get started quickly. The demo datasets are hosted by NOW which can be easily
used to build a search application. There is a large variety of datasets, including images, text, and audio.

If you would like to use your own custom data, you can choose `DocumentArray name`. You will be asked to provide the
DocumentArray ID (or name) or URL of your dataset.

```commandline
? Please enter your DocumentArray name:
```

You can also choose the local folder option to upload your data, in which case
NOW asks for the path to the folder containing your data:

```commandline
? Please enter the path to the local folder:
```

Perhaps your data is stored in an S3 bucket, which is an option NOW also supports. In this case,
NOW asks for the URI to the S3 bucket, as well as the credentials and region thereof.

```commandline
? Please enter the S3 URI to the folder:
? Please enter the AWS access key ID:
? Please enter the AWS secret access key:
? Please enter the AWS region:
```

A final step in loading your data is to choose the fields of your data that you would like to use for search and filter
respectively. You can choose from the fields that are available in your data. For more information on this, see the
[guide to loading your data](https://now.jina.ai/user-guides/input_format/).


**2.** Choose your deployment type.

We recommend cloud deployment to run your app on our servers. Alternatively, you can deploy locally.

```commandline
? Where do you want to deploy your search engine?  (Use arrow keys)
 ❯ ⛅️ Jina Cloud
   📍 Local
```

For local deployment, NOW asks where you want to deploy it and reads your local `.kube/config` and lists all Kubernetes clusters you have access to.
If you don't want to use an existing cluster, you can create a new one locally.

```commandline
? Which cluster do you want to use to deploy your search engine? (Use arrow keys)
    kind-jina-now
    minikube
  ❯ 🐣create new
```


**3.** Follow the links. After NOW finishes processing, you'll see two links:

- The Swagger UI is useful for frontend integration.
- The "playground" lets you run example queries and experiment with your search use case.

```commandline
🚀 Deploy playground and BFF

BFF docs are accessible at:
http://localhost:30090/api/docs

Playground is accessible at:
http://localhost:30080/?host=gateway&search_field=image&data=best-artworks&port=8080
```

Example of the playground.

<img width="350" alt="Screenshot 2022-05-26 at 16 36 49" src="https://user-images.githubusercontent.com/11627845/170511607-3fb810f7-a5aa-47cd-9f70-e6034a96b9fd.png">

Example of the Swagger UI.

<img width="350" alt="Screenshot 2022-05-26 at 16 36 06" src="https://user-images.githubusercontent.com/11627845/170511580-230d1e41-5e14-4623-adb6-3d4b2d400dc9.png">

[More information on using Jina NOW CLI and API](https://now.jina.ai/user-guides/cli_api/)

## Supported modalities (more coming soon)

- 📝 Text
- 🏞 Image
- 🥁 Music
- 🎥 Video (for GIFs)
- 🧊 3D Mesh (coming soon)

[![](https://user-images.githubusercontent.com/11627845/164571632-0e6a6c39-0137-413b-8287-21fc34785665.png)](https://www.youtube.com/watch?v=fdIaLP0ctpo)
</p>
  
## Example datasets

<details><summary>👕 Fashion</summary>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/11627845/157079335-8f36fc73-d826-4c0a-b1f3-ed5d650a1af1.png">
</details>

<details><summary>☢️ Chest X-Ray</summary>
<img src="https://user-images.githubusercontent.com/11627845/157067695-59851a77-5c43-4f68-80c4-403fec850776.png" width="400">
</details>
  
<details><summary>💰 NFT - bored apes</summary>
<img src="https://user-images.githubusercontent.com/11627845/157019002-573cc101-e23b-4020-825c-f37ec66c6ccf.jpeg" width="400">
</details>
  
<details><summary>🖼 Art</summary>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/11627845/157074453-721c0f2d-3f7d-4839-b6ff-bbccbdba2e5f.png">
</details>
  
<details><summary>🚗 Cars</summary>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/11627845/157081047-792df6bd-544d-420c-b180-df824c802e73.png">
</details>
  
<details><summary>🏞 Street view</summary>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/11627845/157087532-46ae36a2-c97f-45d7-9c3e-c624dcf6dc46.png">
</details>

<details><summary>🦆 Birds</summary>
<img width="400" alt="image" src="https://user-images.githubusercontent.com/11627845/157069954-615a5cb6-dda0-4a2f-9442-ea807ad4a8d5.png">
</details>

<!-- end elevator-pitch -->



CHANGELOG.md














































<a name=release-note-0-0-7></a>
## Release Note (`0.0.7`)

> Release time: 2022-05-18 10:16:30



🙇 We'd like to thank all contributors for this new release! In particular,
 Mohammad Kalim Akram,  Florian Hönicke,  Florian Hönicke,  joschkabraun,  Sebastian Lettner,  Roshan Jossy,  makram93,  🙇


### 🆕 New Features

 - [[```cf89bb7b```](https://github.com/jina-ai/now/commit/cf89bb7bab23dd18b1dcf20d68132f30b903da5c)] __-__ survey (#174) (*Florian Hönicke*)
 - [[```b6419278```](https://github.com/jina-ai/now/commit/b64192780f1bf5d0906fcbd5cf19e1b311aba860)] __-__ wolf support (#156) (*Mohammad Kalim Akram*)
 - [[```0ec8cafc```](https://github.com/jina-ai/now/commit/0ec8cafcd9b5d8667e678bcbd3bb8150f0bf1a2d)] __-__ __custom-text-data__: implemented custom text data loading (*joschkabraun*)
 - [[```a5f09e35```](https://github.com/jina-ai/now/commit/a5f09e35651bece1713a945a9b6bf4b3ea4e8d1c)] __-__ add bff using FastAPI (#112) (*Mohammad Kalim Akram*)
 - [[```38555eb9```](https://github.com/jina-ai/now/commit/38555eb907b59eb976303a5645019c42aced5606)] __-__ update logo shown on frontend (#128) (*Mohammad Kalim Akram*)
 - [[```cfd26542```](https://github.com/jina-ai/now/commit/cfd26542ae865058c18d66baebd82d9b6155ac40)] __-__ __dialog__: add install instructions audio (#149) (*Sebastian Lettner*)
 - [[```6e647def```](https://github.com/jina-ai/now/commit/6e647def2148a7417ee279a0d9c0449cb6a85e5e)] __-__ data loading music (#120) (*Sebastian Lettner*)
 - [[```cd690cc2```](https://github.com/jina-ai/now/commit/cd690cc2c8c7d45c7c4f16936bd306c0cd39b625)] __-__ __frontend-cd__: added frontend deployment to ci (#121) (*joschkabraun*)
 - [[```580128f2```](https://github.com/jina-ai/now/commit/580128f24e68e7143ec2895cde3344ac73c00412)] __-__ bump python version (*Florian Hönicke*)
 - [[```1ae58792```](https://github.com/jina-ai/now/commit/1ae58792392969cff043e8f734f1c416f091f850)] __-__ sandbox support (#101) (*Florian Hönicke*)
 - [[```09105a6d```](https://github.com/jina-ai/now/commit/09105a6daec3b9fd41a8a5b5b157acfaf3f2f1c4)] __-__ webcam support (#111) (*Mohammad Kalim Akram*)
 - [[```dbf01964```](https://github.com/jina-ai/now/commit/dbf0196435de272d11c22b092def34cb0ca85776)] __-__ __dialog__: add modality selection (#107) (*Sebastian Lettner*)
 - [[```d194a8cb```](https://github.com/jina-ai/now/commit/d194a8cbe02ec03243b496d91dd56d600119f107)] __-__ no head deployments (#100) (*Florian Hönicke*)
 - [[```86ed814e```](https://github.com/jina-ai/now/commit/86ed814e14664f8f15d3786b5983c0afd637e873)] __-__ add slider for confidence filtering (#95) (*Mohammad Kalim Akram*)
 - [[```086ad7f1```](https://github.com/jina-ai/now/commit/086ad7f163d53da94693650d66c3571107b2c430)] __-__ thumbnail images to (256, 256) (#94) (*joschkabraun*)
 - [[```0aa23727```](https://github.com/jina-ai/now/commit/0aa2372788feab9ca527711a62aa00934255ad57)] __-__ add codecov and cicd (#83) (*Mohammad Kalim Akram*)
 - [[```6b31d08e```](https://github.com/jina-ai/now/commit/6b31d08e8ebcd87cd8e9e469898f30323874be7a)] __-__ most simple indexer (#54) (*Florian Hönicke*)
 - [[```8daee7aa```](https://github.com/jina-ai/now/commit/8daee7aa9f78e864ac132ffdaeb3b1055662f042)] __-__ handle bad cluster (#53) (*Florian Hönicke*)
 - [[```cbd148c5```](https://github.com/jina-ai/now/commit/cbd148c5fdc06bdcebc2792551a0227024422fdd)] __-__ min version (*Florian Hönicke*)

### 🐞 Bug fixes

 - [[```80d3fb77```](https://github.com/jina-ai/now/commit/80d3fb77b3812a5270673bccfbf90ecf108170fa)] __-__ __ci__: add dependency in cd (#175) (*Mohammad Kalim Akram*)
 - [[```c299c343```](https://github.com/jina-ai/now/commit/c299c3434f680d8ff4e265664d941dee07a2e899)] __-__ add logo (*Florian Hönicke*)
 - [[```fe40a97e```](https://github.com/jina-ai/now/commit/fe40a97e8e89d4924275afb0c2f9e3b9a06ead59)] __-__ svg logo (#169) (*Florian Hönicke*)
 - [[```d616bff7```](https://github.com/jina-ai/now/commit/d616bff7d8816fb5e57390db791bb49c7155a5c0)] __-__ workaround first request hangs (#162) (*Florian Hönicke*)
 - [[```50e5791a```](https://github.com/jina-ai/now/commit/50e5791af06329e14c9b4d17ce82f22c5ed981c6)] __-__ check version (#158) (*Florian Hönicke*)
 - [[```37b7bfe7```](https://github.com/jina-ai/now/commit/37b7bfe72487a8290f717113f0b77199592931c6)] __-__ indexer handles limit (#160) (*Florian Hönicke*)
 - [[```0b80514c```](https://github.com/jina-ai/now/commit/0b80514c383337e5782440979340100b28c49586)] __-__ deactivate sandbox (#151) (*Florian Hönicke*)
 - [[```b3797e8a```](https://github.com/jina-ai/now/commit/b3797e8a05fc18c16b4a6816610649099fdf6e74)] __-__ waiting at zero percent (#150) (*Florian Hönicke*)
 - [[```ddbf6e85```](https://github.com/jina-ai/now/commit/ddbf6e851e6b59e26d0993d56de56bb721860929)] __-__ only import pydub in case of music (#146) (*Florian Hönicke*)
 - [[```4c1e9da6```](https://github.com/jina-ai/now/commit/4c1e9da6fd3b46d1aa4a5155ea389382e818c28c)] __-__ modality (*Florian Hönicke*)
 - [[```01897a9f```](https://github.com/jina-ai/now/commit/01897a9f258120f0a28df70f0f94cf356107ae21)] __-__ update frontend version (#108) (*Florian Hönicke*)
 - [[```88ad87c5```](https://github.com/jina-ai/now/commit/88ad87c547ad6a9cba623bd9d1e0d0683a1f516b)] __-__ fixed jina version (#109) (*Florian Hönicke*)
 - [[```db18d384```](https://github.com/jina-ai/now/commit/db18d384e6af1425a4ea72f04a312f96f922bd99)] __-__ __ci__: update workflows without release_token (*makram93*)
 - [[```9994c301```](https://github.com/jina-ai/now/commit/9994c301f64e0320a31c9eae18210f160dcd9fdd)] __-__ fix the url for docs deployment (*makram93*)
 - [[```ce5a8a21```](https://github.com/jina-ai/now/commit/ce5a8a21767285b1e15476c7de1853e5e6772897)] __-__ use command to get context lists (#93) (*Mohammad Kalim Akram*)
 - [[```19ffb34f```](https://github.com/jina-ai/now/commit/19ffb34f15eaf97ed693527a980ba11c26efbd89)] __-__ use cpu only (#92) (*Mohammad Kalim Akram*)

now/constants.py
from __future__ import annotations, print_function, unicode_literals

from docarray.typing import Image, Text, Video

from now.utils import BetterEnum

NOW_GATEWAY_VERSION = '0.0.4-refactor-remove-survey-2'
NOW_PREPROCESSOR_VERSION = '0.0.122-custom-gateway-grpc-secured-53'
NOW_ELASTIC_INDEXER_VERSION = '0.0.147-test-experiment-volume-1'
NOW_AUTOCOMPLETE_VERSION = '0.0.11-refactor-custom-gateway-103'


class Apps(BetterEnum):
    SEARCH_APP = 'search_app'


class DialogStatus(BetterEnum):
    CONTINUE = 'continue'
    BREAK = 'break'
    SKIP = 'skip'


class DatasetTypes(BetterEnum):
    DEMO = 'demo'
    PATH = 'path'
    DOCARRAY = 'docarray'
    S3_BUCKET = 's3_bucket'
    ELASTICSEARCH = 'elasticsearch'


class Models(BetterEnum):
    CLIP_MODEL = 'encoderclip'
    SBERT_MODEL = 'encodersbert'


SUPPORTED_FILE_TYPES = {
    Text: ['txt', 'md'],
    Image: ['jpg', 'jpeg', 'png', 'bmp', 'tiff', 'tif'],
    Video: ['gif'],
}

FILETYPE_TO_MODALITY = {
    filetype: modality
    for modality, filetypes in SUPPORTED_FILE_TYPES.items()
    for filetype in filetypes
}
AVAILABLE_MODALITIES_FOR_SEARCH = [Text, Image, Video]
AVAILABLE_MODALITIES_FOR_FILTER = [Text]
NOT_AVAILABLE_MODALITIES_FOR_FILTER = [Image, Video]

BASE_STORAGE_URL = (
    'https://storage.googleapis.com/jina-fashion-data/data/one-line/datasets'
)

EXTERNAL_CLIP_HOST = 'encoderclip-pretty-javelin-3aceb7f2cd.wolf.jina.ai'

DEFAULT_FLOW_NAME = 'nowapi'
PREFETCH_NR = 10
NUM_FOLDERS_THRESHOLD = 100
SURVEY_LINK = 'https://10sw1tcpld4.typeform.com/to/VTAyYRpR?utm_source=cli'

TAG_INDEXER_DOC_HAS_TEXT = '_indexer_doc_has_text'
ACCESS_PATHS = '@cc'
FLOW_STATUS = 'Serving'
DEMO_NS = 'now-example-{}'

MODALITY_TO_MODELS = {
    Text: [
        {'name': 'Clip', 'value': Models.CLIP_MODEL},
        {'name': 'Sbert', 'value': Models.SBERT_MODEL},
    ],
    Image: [{'name': 'Clip', 'value': Models.CLIP_MODEL}],
    Video: [{'name': 'Clip', 'value': Models.CLIP_MODEL}],
}

NOWGATEWAY_BFF_PORT = 8080
